xarray/core/dataset.py

```diff
@@ -7007,5 +7008,235 @@ def argmax(self, dim=None, axis=None, **kwargs):
+
+        See also
+        --------
+        Dataset.polyfit
+        scipy.optimize.curve_fit
+        """
+        from scipy.optimize import curve_fit
+
+        if p0 is None:
+            p0 = {}
+        if bounds is None:
+            bounds = {}
+        if kwargs is None:
+            kwargs = {}
+
+        if not reduce_dim:
+            reduce_dims = []
+        elif isinstance(reduce_dim, str) or not isinstance(reduce_dim, Iterable):
+            reduce_dims = [reduce_dim]
+        else:
+            reduce_dims = list(reduce_dim)
+
+        if (
+            isinstance(coords, str)
+            or isinstance(coords, xr.DataArray)
+            or not isinstance(coords, Iterable)
+        ):
+            coords = [coords]
+        coords_ = [self[coord] if isinstance(coord, str) else coord for coord in coords]
+
+        # Determine whether any coords are dims on self
+        for coord in coords_:
+            reduce_dims += [c for c in self.dims if coord.equals(self[c])]
+        reduce_dims = list(set(reduce_dims))
+        preserved_dims = list(set([dim for dim in self.dims]) - set(reduce_dims))
+        if not reduce_dims:
+            raise ValueError(
+                "No arguments to `coords` were identified as a dimension on the calling "
+                "object, and no dims were supplied to `reduce_dim`. This would result "
+                "in fitting on scalar data."
+            )
+
+        # Broadcast all coords with each other
+        coords_ = xr.broadcast(*coords_)
+        coords_ = [
+            coord.broadcast_like(self, exclude=preserved_dims) for coord in coords_
+        ]
+
+        def _initialize_feasible(lb, ub):
+            # Mimics functionality of scipy.optimize.minpack._initialize_feasible
+            lb_finite = np.isfinite(lb)
+            ub_finite = np.isfinite(ub)
+            p0 = (
+                0.5 * (lb + ub) * (lb_finite & ub_finite)
+                + (lb + 1) * (lb_finite & ~ub_finite)
+                + (ub - 1) * (~lb_finite & ub_finite)
+            )
+            return p0
+
+        # Get p0 and bounds
+        # Priority: 1) passed args 2) func signature 3) scipy defaults
+        try:
+            func_args = inspect.signature(func).parameters
+        except ValueError:
+            func_args = {}
+            if not param_names:
+                raise ValueError(
+                    "Unable to inspect `func` signature, and `param_names` was not provided."
+                )
+        if param_names:
+            params = param_names
+        else:
+            params = list(func_args)[1:]
+            if any(
+                [
+                    (p.kind in [p.VAR_POSITIONAL, p.VAR_KEYWORD])
+                    for p in func_args.values()
+                ]
+            ):
+                raise ValueError(
+                    "`param_names` must be provided because `func` takes variable length arguments."
+                )
+
+        n_params = len(params)
```

Comment:
Sounds like a good idea. Where should these live?

Reply:
Ok. I guess they could either be module-level private functions at the top of `dataset.py` (but I can't imagine them being used elsewhere), or we could put them within the `Dataset` class with a `@staticmethod` decorator (this also seems a bit funky). 

Refactoring things out into `fitting.py` sounds like a good long term solution either way.

Reply:
private functions sounds fine to me.

I didn't realize this was in `dataset.py` Another option is to move it to `computation.py`.

Reply:
`polyfit` is in `dataset.py`, and I do think it makes sense to call it as a dataset method, since a common use case would be just fitting a function along a dimension. But factoring out most of the code to functions in `computation.py` or a new `fitting.py` would help keep things organized.

Reply:
I think we add private methods for now and refactor later.